File: swift-sdks/Sources/SuiSwiftSDK/DeepBookV3/DeepBookTPSL.swift
import Foundation

// Take Profit / Stop Loss Operations for Swift SDK
public class DeepBookTPSL {
    weak var client: DeepBookV3Client?

    public init(client: DeepBookV3Client) {
        self.client = client
    }

    // MARK: - Conditional Orders

    public func addConditionalOrder(
        sender: String,
        marginManagerKey: String,
        poolKey: String,
        targetPrice: UInt64,
        quantity: UInt64,
        isBid: Bool,
        priceDirection: Bool
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: (client?.config.deepbookPackage ?? "") + "::margin_manager::add_conditional_order",
            arguments: [
                [
                    "marginManagerKey": marginManagerKey,
                    "poolKey": poolKey,
                    "targetPrice": targetPrice,
                    "quantity": quantity,
                    "isBid": isBid,
                    "priceDirection": priceDirection
                ]
            ]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    public func cancelConditionalOrder(
        sender: String,
        marginManagerKey: String,
        orderId: String
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: (client?.config.deepbookPackage ?? "") + "::margin_manager::cancel_conditional_order",
            arguments: [
                [
                    "marginManagerKey": marginManagerKey,
                    "orderId": orderId
                ]
            ]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    public func cancelAllConditionalOrders(
        sender: String,
        marginManagerKey: String
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: (client?.config.deepbookPackage ?? "") + "::margin_manager::cancel_all_conditional_orders",
            arguments: [["marginManagerKey": marginManagerKey]]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    public func executeConditionalOrders(
        sender: String,
        poolKey: String
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: (client?.config.deepbookPackage ?? "") + "::margin_manager::execute_conditional_orders",
            arguments: [["poolKey": poolKey]]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    // MARK: - Take Profit

    public func setTakeProfit(
        sender: String,
        marginManagerKey: String,
        targetPrice: UInt64,
        quantity: UInt64
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: (client?.config.deepbookPackage ?? "") + "::margin_manager::set_take_profit",
            arguments: [
                [
                    "marginManagerKey": marginManagerKey,
                    "targetPrice": targetPrice,
                    "quantity": quantity
                ]
            ]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    public func cancelTakeProfit(
        sender: String,
        marginManagerKey: String
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: (client?.config.deepbookPackage ?? "") + "::margin_manager::cancel_take_profit",
            arguments: [["marginManagerKey": marginManagerKey]]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    // MARK: - Stop Loss

    public func setStopLoss(
        sender: String,
        marginManagerKey: String,
        targetPrice: UInt64,
        quantity: UInt64
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: (client?.config.deepbookPackage ?? "") + "::margin_manager::set_stop_loss",
            arguments: [
                [
                    "marginManagerKey": marginManagerKey,
                    "targetPrice": targetPrice,
                    "quantity": quantity
                ]
            ]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    public func cancelStopLoss(
        sender: String,
        marginManagerKey: String
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: (client?.config.deepbookPackage ?? "") + "::margin_manager::cancel_stop_loss",
            arguments: [["marginManagerKey": marginManagerKey]]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    // MARK: - Helper Methods

    private func call(_ method: String, params params: [Any?] = []) async throws -> [String: Any] {
        guard let client = client else {
            throw DeepBookError.invalidResponse
        }
        return try await client.call(method, params: params)
    }

    private func buildTransaction(target: String, arguments: [[String: Any]]) -> [String: Any] {
        var transactions: [[String: Any]] = []
        
        for args in arguments {
            var moveCall: [String: Any] = [:]
            moveCall["kind"] = "moveCall"
            moveCall["target"] = target
            moveCall["arguments"] = args
            transactions.append(moveCall)
        }

        return [
            "kind": "programmableTransaction",
            "inputs": [],
            "transactions": transactions
        ]
    }
}

File: java-sdks/src/main/java/com/suisdks/sui/deepbook_v3/DeepBookTPSL.java
package com.suisdks.sui.deepbook_v3;

import java.math.BigInteger;
import java.util.*;

public class DeepBookTPSL {
    private final DeepBookV3Client client;

    public DeepBookTPSL(DeepBookV3Client client) {
        this.client = client;
    }

    // ==================== Conditional Orders (TPSL) ====================

    public Map<String, Object> addConditionalOrder(String sender,
            String marginManagerKey, String poolKey,
            boolean triggerBelowPrice, String triggerPrice,
            String orderType, String orderSide, BigInteger quantity,
            BigInteger price, String clientOrderId) {
        Map<String, Object> txData = new HashMap<>();
        txData.put("kind", "programmableTransaction");

        List<Map<String, Object>> transactions = new ArrayList<>();
        
        // First: Set conditional order
        Map<String, Object> setOrder = new HashMap<>();
        setOrder.put("kind", "moveCall");
        setOrder.put("target", client.config().marginPackage() + "::margin_tpsl::set_conditional_order");
        Map<String, Object> orderArgs = new HashMap<>();
        orderArgs.put("marginManagerKey", marginManagerKey);
        orderArgs.put("triggerBelowPrice", triggerBelowPrice);
        orderArgs.put("triggerPrice", triggerPrice);
        orderArgs.put("orderType", orderType);
        orderArgs.put("orderSide", orderSide);
        orderArgs.put("quantity", quantity.toString());
        orderArgs.put("price", price.toString());
        orderArgs.put("clientOrderId", clientOrderId);
        setOrder.put("arguments", List.of(orderArgs));
        transactions.add(setOrder);

        txData.put("inputs", List.of());
        txData.put("transactions", transactions);

        return client.callRpc("suix_devInspectTransactionBlock", List.of(sender, txData));
    }

    public Map<String, Object> cancelConditionalOrder(String sender, String marginManagerKey,
            String conditionalOrderId) {
        Map<String, Object> txData = new HashMap<>();
        txData.put("kind", "programmableTransaction");

        Map<String, Object> moveCall = new HashMap<>();
        moveCall.put("kind", "moveCall");
        moveCall.put("target", client.config().marginPackage() + "::margin_tpsl::cancel_conditional_order");
        
        Map<String, Object> args = new HashMap<>();
        args.put("marginManagerKey", marginManagerKey);
        args.put("conditionalOrderId", conditionalOrderId);
        moveCall.put("arguments", List.of(args));
        
        txData.put("inputs", List.of());
        txData.put("transactions", List.of(moveCall));

        return client.callRpc("suix_devInspectTransactionBlock", List.of(sender, txData));
    }

    public Map<String, Object> cancelAllConditionalOrders(String sender, String marginManagerKey) {
        Map<String, Object> txData = new HashMap<>();
        txData.put("kind", "programmableTransaction");

        Map<String, Object> moveCall = new HashMap<>();
        moveCall.put("kind", "moveCall");
        moveCall.put("target", client.config().marginPackage() + "::margin_tpsl::cancel_all_conditional_orders");
        
        Map<String, Object> args = new HashMap<>();
        args.put("marginManagerKey", marginManagerKey);
        moveCall.put("arguments", List.of(args));
        
        txData.put("inputs", List.of());
        txData.put("transactions", List.of(moveCall));

        return client.callRpc("suix_devInspectTransactionBlock", List.of(sender, txData));
    }

    public Map<String, Object> executeConditionalOrders(String managerAddress, String poolKey,
        int maxOrders, String gasPrice) {
        Map<String, Object> txData = new HashMap<>();
        txData.put("kind", "programmableTransaction");

        Map<String, Object> moveCall = new HashMap<>();
        moveCall.put("kind", "moveCall");
        moveCall.put("target", client.config().marginPackage() + "::margin_tpsl::execute_conditional_orders");
        
        Map<String, Object> data = new HashMap<>();
        data.put("marginManagerAddress", managerAddress);
        data.put("poolKey", poolKey);
        data.put("maxOrders", maxOrders);
        
        Map<String, Object> params = new HashMap<>();
        params.put("data", data);
        params.put("gasPrice", gasPrice);
        
        // Add Move call input for maxOrders and gasPrice
        List<Map<String, Object>> transactions = new ArrayList<>();
        transactions.add(moveCall);
        txData.put("transactions", transactions);
        txData.put("inputs", List.of());

        return client.callRpc("suix_devInspectTransactionBlock", List.of(
            client.config().address(), txData
        ));
    }

    public Map<String, Object> getConditionalOrders(String poolKey, String marginManagerId) {
        return client.callRpc("suix_getDynamicField", List.of(
            poolKey + "::margin_manager::conditional_orders",
            Map.of(
                "type", "string",
                "key", marginManagerId
            )
        ));
    }

    public Map<String, Object> getConditionalOrder(String poolKey, String marginManagerId,
            String conditionalOrderId) {
        return client.callRpc("suix_getDynamicField", List.of(
            poolKey + "::margin_manager::conditional_order_" + conditionalOrderId,
            Map.of("type", "u64")
        ));
    }

    public String lowestTriggerAbovePrice(String poolKey, String marginManagerId) {
        Map<String, Object> state = getMarginManagerState(marginManagerId);
        Map<String, Object> data = (Map<String, Object>) state.get("data");
        
        if (data != null && data.containsKey("fields")) {
            Map<String, Object> fields = (Map<String, Object>) data.get("fields");
            return (String) fields.getOrDefault("lowest_trigger_above_price", "0");
        }
        
        return "0";
    }

    public String highestTriggerBelowPrice(String poolKey, String marginManagerId) {
        Map<String, Object> state = getMarginManagerState(marginManagerId);
        Map<String, Object> data = (Map<String, Object>) state.get("data");
        
        if (data != null && data.containsKey("fields")) {
            Map<String, Object> fields = (Map<String, Object>) data.get("fields");
            return (String) fields.getOrDefault("highest_trigger_below_price", "0");
        }
        
        return "0";
    }

    private Map<String, Object> getMarginManagerState(Object marginManagerId) {
        return client.callRpc("sui_getObject", Map.of(
            "objectId", marginManagerId,
            "options", Map.of("showContent", true)
        ));
    }
}